# REDES NEURAIS ARTIFICIAIS
## Evaluate Improve e Tunning

# Parte 1 - Pré-processamento dos Dados

### Importar as libs

In [31]:
import numpy as np
import pandas as pd
import tensorflow as tf

In [32]:
tf.__version__

'2.8.0'

### Importar o dataset

In [33]:
dataset = pd.read_csv('Churn_Modelling.csv')
dataset.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [34]:
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values
print("X >>\n",X)
print("y >>\n",y)

X >>
 [[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]
y >>
 [1 0 1 ... 1 1 0]


### Transformando os dados categóricos


In [35]:
# Label Encoding the "Gender" column
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])
print(X)
# One Hot Encoding the "Geography" column
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]
[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


# Dividindo o dataset em conjunto de treinamento e conjunto de teste

In [36]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling

In [37]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [38]:
np.shape(X_train)

(8000, 12)

---
# Parte 2 -Vamos construir uma ANN!


### Initializing the ANN

In [9]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [10]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [11]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

In [12]:
ann.add(tf.keras.layers.Dense(units=1, activation='tanh'))

## Part 3 - Training the ANN

### Compiling the ANN

In [13]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

In [14]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 1s 716us/step - loss: 0.6999 - accuracy: 0.6961
Epoch 2/100
250/250 [==============================] - 0s 704us/step - loss: 0.5317 - accuracy: 0.7720
Epoch 3/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4891 - accuracy: 0.7918
Epoch 4/100
250/250 [==============================] - 0s 755us/step - loss: 0.4669 - accuracy: 0.8009
Epoch 5/100
250/250 [==============================] - 0s 879us/step - loss: 0.4523 - accuracy: 0.8043
Epoch 6/100
250/250 [==============================] - 0s 951us/step - loss: 0.4428 - accuracy: 0.8073
Epoch 7/100
250/250 [==============================] - 0s 720us/step - loss: 0.4361 - accuracy: 0.8114
Epoch 8/100
250/250 [==============================] - 0s 866us/step - loss: 0.4314 - accuracy: 0.8135
Epoch 9/100
250/250 [==============================] - 0s 827us/step - loss: 0.4277 - accuracy: 0.8149
Epoch 10/100
250/250 [==============================] - 0s 856us/step - los

# Parte 3 - Fazendo predições e avaliando o modelo

## Prevendo os resultados com o conjunto de testes

In [15]:
y_pred = ann.predict(X_test)
y_pred

array([[0.19997032],
       [0.31730822],
       [0.15073824],
       ...,
       [0.08161885],
       [0.15665542],
       [0.14460936]], dtype=float32)

In [16]:
y_pred = (y_pred > 0.5)
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

# Criando uma Confusion Matrix

In [17]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1511   84]
 [ 193  212]]


---
# Avaliando o Modelo

In [18]:
# Evaluating the ANN
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [19]:
def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 12))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'tanh'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

In [20]:
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 10)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10, n_jobs = -1)
mean = accuracies.mean()
variance = accuracies.std()

C:\Users\luanm\AppData\Local\Temp\ipykernel_8664\1297209367.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 10)


In [21]:
print(mean,variance)

0.8350000023841858 0.021491280032177488


# Melhorando o modelo

In [22]:
# Improving the ANN
# Dropout Regularization to reduce overfitting if needed
from tensorflow.keras.layers import Dropout

def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 12))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'tanh'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

In [23]:
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 10)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10, n_jobs = -1)
mean = accuracies.mean()
variance = accuracies.std()

C:\Users\luanm\AppData\Local\Temp\ipykernel_8664\1297209367.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 10)


In [24]:
print(mean,variance)

0.8197500050067902 0.01914418679554296


# Tuning the ANN

In [25]:
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

def build_classifier(optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 12))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'tanh'))
    classifier.compile(optimizer = optimizer, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

classifier = KerasClassifier(build_fn = build_classifier)

parameters = {
    'batch_size': [10, 25, 32],
    'epochs': [50, 100],
    'optimizer': ['adam', 'rmsprop']
}

grid_search = GridSearchCV(estimator = classifier,
                           scoring = 'accuracy',
                           param_grid = parameters,
                           cv = 10)

C:\Users\luanm\AppData\Local\Temp\ipykernel_8664\1850979573.py:16: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  classifier = KerasClassifier(build_fn = build_classifier)


In [26]:
%%time

grid_search = grid_search.fit(X_train, y_train)

Epoch 1/50
720/720 [==============================] - 3s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 2/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 3/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 4/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 5/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 6/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 7/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 8/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 9/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971
Epoch 10/50
720/720 [==============================] - 2s 3ms/step - loss: 3.1300 - accuracy: 0.7971

In [27]:
print('Best params:\n',grid_search.best_params_,'\nBest accuracy:\n',grid_search.best_score_)

Best params:
 {'batch_size': 25, 'epochs': 100, 'optimizer': 'rmsprop'} 
Best accuracy:
 0.853125


# Halving Grid Search

In [28]:
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.experimental import enable_halving_search_cv
from sklearn.model_selection import HalvingGridSearchCV

def build_classifier(optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 12))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dropout(rate = 0.1))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'tanh'))
    classifier.compile(optimizer = optimizer, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

classifier = KerasClassifier(build_fn = build_classifier)

parameters = {
    'batch_size': [10, 25, 32],
    'epochs': [50, 100],
    'optimizer': ['adam', 'rmsprop']
}

halving_cv = HalvingGridSearchCV(estimator = classifier,
                                  param_grid = parameters,
                                  scoring="accuracy", 
                                  n_jobs=-1, 
                                  min_resources="exhaust", 
                                  factor=3,
                                  cv = 10)

C:\Users\luanm\AppData\Local\Temp\ipykernel_8664\78126798.py:17: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  classifier = KerasClassifier(build_fn = build_classifier)


In [29]:
%%time
halving_cv = halving_cv.fit(X_train, y_train)

Epoch 1/100
320/320 [==============================] - 1s 2ms/step - loss: 0.5409 - accuracy: 0.7972
Epoch 2/100
320/320 [==============================] - 1s 2ms/step - loss: 0.4234 - accuracy: 0.8024
Epoch 3/100
320/320 [==============================] - 1s 2ms/step - loss: 0.4168 - accuracy: 0.8106
Epoch 4/100
320/320 [==============================] - 1s 2ms/step - loss: 0.4101 - accuracy: 0.8106
Epoch 5/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3986 - accuracy: 0.8170
Epoch 6/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3964 - accuracy: 0.8199
Epoch 7/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3905 - accuracy: 0.8256
Epoch 8/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3856 - accuracy: 0.8316
Epoch 9/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3783 - accuracy: 0.8375
Epoch 10/100
320/320 [==============================] - 1s 2ms/step - loss: 0.3786 - accura

In [30]:
print('Best params:\n',halving_cv.best_params_,'\nBest accuracy:\n',halving_cv.best_score_)

Best params:
 {'batch_size': 25, 'epochs': 100, 'optimizer': 'rmsprop'} 
Best accuracy:
 0.8445556946182728


Média: 0.8197500050067902
Acurácia: 0.8445556946182728
Variação: 0.01914418679554296

# FIM

> Professor Diego Dorgam  
> [@diegodorgam](https://twitter.com/diegodorgam)